### PR TITLE
Public Dashboards: Query Caching

### PR DESCRIPTION
**Pull request by owensmallwood**
**What this does**
This is an enterprise-only feature. Queries for panels that come from a public dashboard will be cached the same way queries from regular dashboard panels do. [Enterprise PR](https://github.com/grafana/grafana-enterprise/pull/3518)

**How to test**
- Enable public dashboards and query caching in your ini
`publicDashboards = true`
`caching.queries = true`
- Enable caching for the datasource(s) that your public dashboard uses
- Check your network tab for the `query` request when running a public dashboard. After you refresh the page, the `X-Cache` header should show as a `HIT`

### CODE REVIEW

**Review by jalevin — APPROVED** [review]

File: pkg/services/dashboards/service/dashboard_public.go [hunk 1]
```diff
@@ -143,6 +144,26 @@ func (dr *DashboardServiceImpl) BuildPublicDashboardMetricRequest(ctx context.Co
 	}, nil
 }
 
+// BuildAnonymousUser creates a user with permissions to read from all datasources used in the dashboard
+func (dr *DashboardServiceImpl) BuildAnonymousUser(ctx context.Context, dashboard *models.Dashboard) (*models.SignedInUser, error) {
```
[the remark this answers is not in the document]
Comment (owensmallwood) [reply pkg/services/dashboards/service/dashboard_public.go:148]: this gets called in enterprise in `query_middleware.go`

File: pkg/models/dashboard_queries_test.go [hunk 2]
```diff
@@ -56,6 +56,57 @@ const (
   "schemaVersion": 35
 }`
 
+	dashboardWithDuplicateDatasources = `
```
[the remark this answers is not in the document]
Comment (owensmallwood) [reply pkg/models/dashboard_queries_test.go:59]: Ya sure I'll add that

File: pkg/models/dashboard_queries.go [hunk 3]
```diff
@@ -4,7 +4,23 @@ import (
...
+		if _, ok := exists[uid]; !ok {
+			datasourceUids = append(datasourceUids, uid)
+			exists[uid] = true
+		}
+	}
+
+	return datasourceUids
+}
+
+func GroupQueriesByPanelId(dashboard *simplejson.Json) map[int64][]*simplejson.Json {
```
[the remark this answers is not in the document]
Comment (owensmallwood) [reply pkg/models/dashboard_queries.go:23]: It seemed more descriptive of what it does (gets the dashboard queries and groups them by panel id), and it was similar to the function below.

File: pkg/models/dashboard_queries.go [hunk 4]
```diff
@@ -4,7 +4,23 @@ import (
...
+	exists := map[string]bool{}
+
+	for _, panelObj := range dashboard.Get("panels").MustArray() {
+		panel := simplejson.NewFromAny(panelObj)
+		uid := panel.Get("datasource").Get("uid").MustString()
+		if _, ok := exists[uid]; !ok {
+			datasourceUids = append(datasourceUids, uid)
+			exists[uid] = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (evictorero) [reply pkg/models/dashboard_queries.go:18]: Me neither! but we can't use it until we upgrade to 1.18

**Review by evictorero — APPROVED** [review]

**Pull request merged**